libraries/apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/FetchPolicyInterceptors.kt

```diff
@@ -33,83 +33,72 @@ val CacheOnlyInterceptor = object : ApolloInterceptor {
   }
 }
 
+/**
+ * An interceptor that emits the response(s) from the network only.
+ */
 val NetworkOnlyInterceptor = object : ApolloInterceptor {
   override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
     return chain.proceed(request)
   }
 }
 
 /**
- * An interceptor that goes to cache first and then to the network if it fails
+ * An interceptor that emits the response from the cache first, and if there was a cache miss, emits the response(s) from the network.
  */
 val CacheFirstInterceptor = object : ApolloInterceptor {
   override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
     return flow {
-      var cacheException: ApolloException? = null
-      var networkException: ApolloException? = null
-
       val cacheResponse = chain.proceed(
           request = request
               .newBuilder()
               .fetchFromCache(true)
               .build()
-      ).catch { throwable ->
-        if (throwable is ApolloException) {
-          cacheException = throwable
-        } else {
-          throw throwable
-        }
-      }.singleOrNull()
-
-      if (cacheResponse != null) {
-        emit(cacheResponse)
+      ).single()
+      emit(cacheResponse.newBuilder().isLast(cacheResponse.exception == null).build())
+      if (cacheResponse.exception == null) {
         return@flow
       }
 
       val networkResponses = chain.proceed(
           request = request
-      ).catch {
-        if (it is ApolloException) {
-          networkException = it
-        } else {
-          throw it
-        }
-      }.map { response ->
+      ).map { response ->
         response.newBuilder()
             .cacheInfo(
                 response.cacheInfo!!
                     .newBuilder()
-                    .cacheMissException(cacheException as? CacheMissException)
+                    .networkException(response.exception)
+                    .cacheMissException(cacheResponse.exception as? CacheMissException)
```

Comment:
Is that still required now that the exceptions are set in `FetchPolicyRouterInterceptor`?